tool/tsh/common/hardware_key_test.go

```diff
@@ -0,0 +1,173 @@
+)
+
+// TestHardwareKeyLogin tests Hardware Key login and relogin flows.
+func TestHardwareKeyLogin(t *testing.T) {
+	ctx := context.Background()
+
+	testModules := &modules.TestModules{TestBuildType: modules.BuildEnterprise}
+	modules.SetTestModules(t, testModules)
+
+	connector := mockConnector(t)
+
+	alice, err := types.NewUser("alice@example.com")
+	require.NoError(t, err)
+	aliceRole, err := types.NewRole("alice", types.RoleSpecV6{})
+	require.NoError(t, err)
+	alice.SetRoles([]string{aliceRole.GetName()})
+
+	testServer := testserver.MakeTestServer(t, testserver.WithBootstrap(connector, alice, aliceRole), func(o *testserver.TestServersOpts) {
+		o.ConfigFuncs = append(o.ConfigFuncs, func(cfg *servicecfg.Config) {
+			// TODO (Joerger): This test fails to propagate hardware key policy errors from Proxy SSH connections
+			// for unknown reasons unless Multiplex mode is on. I could not reproduce these errors on a live
+			// cluster, so the issue likely lies with the test server setup. Perhaps the test certs generated
+			// are not 1-to-1 with live certs.
+			cfg.Auth.NetworkingConfig.SetProxyListenerMode(types.ProxyListenerMode_Multiplex)
+		})
+	})
+	authServer := testServer.GetAuthServer()
+
+	proxyAddr, err := testServer.ProxyWebAddr()
+	require.NoError(t, err)
+
+	// mock SSO login and count the number of login attempts.
+	var loginCount int
+	setMockSSOLogin := func(cf *CLIConf) error {
+		loginCount = 0
+		mockSSOLogin := mockSSOLogin(t, authServer, alice)
+		cf.MockSSOLogin = func(ctx context.Context, connectorID string, priv *keys.PrivateKey, protocol string) (*auth.SSHLoginResponse, error) {
+			fmt.Println(priv.GetPrivateKeyPolicy())
+			loginCount++
+			return mockSSOLogin(ctx, connectorID, priv, protocol)
+		}
+		cf.AuthConnector = connector.GetName()
+		return nil
+	}
+
+	t.Run("cap", func(t *testing.T) {
+		setRequireMFAType := func(t *testing.T, requireMFAType types.RequireMFAType) {
+			// Set require MFA type in the cluster auth preference.
+			authPref, err := authServer.UpsertAuthPreference(ctx, &types.AuthPreferenceV2{
+				Spec: types.AuthPreferenceSpecV2{
+					RequireMFAType: requireMFAType,
+				},
+			})
+			require.NoError(t, err)
+
+			// Set MockAttestationData to attest the expected key policy. This will
+			// apply to the next (re)login.
+			testModules.MockAttestationData = &keys.AttestationData{
+				PrivateKeyPolicy: authPref.GetPrivateKeyPolicy(),
+			}
+		}
+
+		t.Cleanup(func() {
+			setRequireMFAType(t, types.RequireMFAType_OFF)
+		})
+
+		// login should use the private key policy reported by the proxy without
+		// needing to retry hardware key login.
+		setRequireMFAType(t, types.RequireMFAType_HARDWARE_KEY_TOUCH)
+		tmpHomePath := t.TempDir()
+		err = Run(context.Background(), []string{
+			"login",
+			"--debug",
+			"--insecure",
+			"--proxy", proxyAddr.String(),
+		}, setHomePath(tmpHomePath), setMockSSOLogin)
+		require.NoError(t, err)
+		require.Equal(t, 1, loginCount)
```

Comment:
```suggestion
		assert.Equal(t, 1, loginCount)
```

Maybe?

A custom message would help too.